### PR TITLE
[WIP] masked_latent_inputs for inpainting pipeline

### PR DESCRIPTION
**Pull request by yiyixuxu**
This PR adds a `masked_image_latents` argument to the inpainting pipeline! 

When the user passes a mask latent (i.e`masked_image_latents`), we also accept image latent and preprocessed masks as inputs.

Currently not working for example 1 described here https://github.com/huggingface/diffusers/issues/4815, i.e. it won't work when user passes only image latents without mask latents. I'm not sure how I should approach this, see https://github.com/huggingface/diffusers/issues/4815#issuecomment-1696428248

example2 from https://github.com/huggingface/diffusers/issues/4815 now working

```python
from diffusers import AutoPipelineForInpainting
from diffusers.utils import load_image
import torch
torch_device = "cuda"

pipe = AutoPipelineForInpainting.from_pretrained("runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, variant="fp16", use_safetensors=True)
pipe = pipe.to(torch_device)

img_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo.png"
mask_url = "https://raw.githubusercontent.com/CompVis/latent-diffusion/main/data/inpainting_examples/overture-creations-5sI6fQgYIuo_mask.png"

init_image = load_image(img_url)
init_mask_image = load_image(mask_url)

prompt = "dslr photography of an empty bench, high quality"
generator = torch.Generator(device="cuda").manual_seed(0)

height, width = 512, 512
image= pipe.image_processor.preprocess(image).to(torch.float16).to(pipe.device)
mask = pipe.mask_processor.preprocess(mask_image).to(torch.float16).to(pipe.device)

masked_image = image * (mask < 0.5)

image = pipe.vae.encode(image).latent_dist.sample() * pipe.vae.config.scaling_factor
mask_latents = pipe.vae.encode(masked_image).latent_dist.sample() * pipe.vae.config.scaling_factor
mask = torch.nn.functional.interpolate(mask, size=(height // pipe.vae_scale_factor, width // pipe.vae_scale_factor))
print(f" mask: {mask.shape}")

image = pipe(
    prompt=prompt,
    image=image,
    mask_image=mask,
    masked_image_latents=mask_latents, # this one is new
    guidance_scale=8.0,
    num_inference_steps=20,
    generator=generator,
).images[0]

image.save(f"inpaint_out.png")
```
yiyi_test_1_out

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Great job! Looks good to me :-) 

Could we:
- 1. Add some fast tests to all pipelines ensuring that passing mask, image latents and normal mask, image leads to the same result?
- 2. Could we also add this logic to SDXL inpainting?

File: tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py [hunk 1]
```diff
@@ -252,6 +252,43 @@ def test_stable_diffusion_inpaint_strength_zero_test(self):
...
+        inputs = self.get_dummy_inputs(device)
+        image = sd_pipe.image_processor.preprocess(inputs["image"]).to(sd_pipe.device)
+        mask = sd_pipe.mask_processor.preprocess(inputs["mask_image"]).to(sd_pipe.device)
+        masked_image = image * (mask < 0.5)
+
+        generator = torch.Generator(device=device).manual_seed(0)
+        image_latents = (
+            sd_pipe.vae.encode(image).latent_dist.sample(generator=generator) * sd_pipe.vae.config.scaling_factor
+        )
+        torch.randn((1, 4, 32, 32), generator=generator)
```
Comment (yiyixuxu) [inline tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py:277]: @patrickvonplaten 

Because the `vae.encode()` is a nondeterministic operation, to completely match the result, I had to force the generator to be used in same sequence in both cases, i.e. encode_initial_image -> create noise to add to it -> encode masked_image. For example, here I had to just use it for the sake of using it just so that the `mask_latents` we created here matches what's in the pipeline. 

I feel pretty stupid doing it this way :pleading_face: is there a better method? also, I think this test will easily fail if we make any changes to the pipeline that change the order in how generator used

Comment (patrickvonplaten) [reply tests/pipelines/stable_diffusion/test_stable_diffusion_inpaint.py:277]: Totally fine for me ! Nice find with the `generator` class & how it works :-)

**Comment by yiyixuxu** [comment]
@patrickvonplaten can I have another look? I updated sdxl too

**Review by patrickvonplaten — APPROVED** [review]
Very nice! Good to merge for me
